atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -163,29 +168,52 @@ protected boolean shouldStopRetrying(int numTimesFailed) {
 
     @Override
     public TransactionAndImmutableTsLock setupRunTaskWithConditionThrowOnConflict(PreCommitCondition condition) {
-        StartIdentifiedAtlasDbTransactionResponse transactionResponse
-                = timelockService.startIdentifiedAtlasDbTransaction();
+        return Iterables.getOnlyElement(setupRunTaskBatchWithConditionThrowOnConflict(ImmutableList.of(condition)));
+    }
+
+    @Override
+    public List<TransactionAndImmutableTsLock> setupRunTaskBatchWithConditionThrowOnConflict(
+            List<PreCommitCondition> conditions) {
+        List<StartIdentifiedAtlasDbTransactionResponse> responses =
+                timelockService.startIdentifiedAtlasDbTransactionBatch(conditions.size());
         try {
-            LockToken immutableTsLock = transactionResponse.immutableTimestamp().getLock();
-            long immutableTs = transactionResponse.immutableTimestamp().getImmutableTimestamp();
-            recordImmutableTimestamp(immutableTs);
-
-            cleaner.punch(transactionResponse.startTimestampAndPartition().timestamp());
-            Supplier<Long> startTimestampSupplier = Suppliers.ofInstance(
-                    transactionResponse.startTimestampAndPartition().timestamp());
-
-            Transaction transaction = createTransaction(
-                    immutableTs,
-                    startTimestampSupplier,
-                    immutableTsLock,
-                    condition);
-            return TransactionAndImmutableTsLock.of(transaction, immutableTsLock);
-        } catch (Throwable e) {
-            timelockService.tryUnlock(ImmutableSet.of(transactionResponse.immutableTimestamp().getLock()));
-            throw Throwables.rewrapAndThrowUncheckedException(e);
+            return wrapResponseBatch(conditions, responses);
+        } catch (Throwable t) {
+            timelockService.tryUnlock(
+                    responses.stream()
+                            .map(response -> response.immutableTimestamp().getLock())
+                            .collect(Collectors.toSet()));
+            throw Throwables.rewrapAndThrowUncheckedException(t);
```

Comment:
The interesting thing here is that I've moved away from doing the approach requiring a specialised object, and I believe the reason it is OK to do so is:
1. The batch will either be in one piece from `TransactionStarter` or it won't (and therefore will have thrown itself);
2. If we fail mid-batch, we still unlock all of them - and there is only a single call, so we don't have to handle cases where the closing itself fails midway, because it either fails or succeeds.